app/src/main/java/edu/wisc/ece/pinpoint/pages/feed/FeedAdapter.java

```diff
@@ -0,0 +1,154 @@
+        holder.item.setOnClickListener(view -> {
+            if (type == ActivityItem.ActivityType.DROP || type == ActivityItem.ActivityType.FIND ||
+                    type == ActivityItem.ActivityType.COMMENT) {
+                if (firebase.getCachedFoundPinMetadata()
+                        .contains(id) || firebase.getCachedDroppedPinMetadata().contains(id))
+                    navController.navigate(NavigationDirections.pinView(id));
+                else
+                    Toast.makeText(parentContext, R.string.undiscovered_pin_locked,
+                                    Toast.LENGTH_SHORT).show();
+
+            } else if (type == ActivityItem.ActivityType.FOLLOW) {
+                navController.navigate(NavigationDirections.profile().setUid(id));
+            }
+        });
+        User cachedAuthor = firebase.getCachedUser(author);
+        if (cachedAuthor != null) {
+            // put author data in feed item
+            setContents(cachedAuthor, type, id, holder, isCurrentUser);
+        } else {
+            // Since only using author for profile pic & username, only fetch if not cached
+            firebase.fetchUser(author)
+                    .addOnCompleteListener(task -> setContents(task.getResult(), type, id, holder, isCurrentUser));
+        }
+    }
+
+    private void setContents(User author, ActivityItem.ActivityType type, String id, @NonNull FeedViewHolder holder, boolean isCurrentUser){
+        author.loadProfilePic(holder.image, fragment);
+        // Detect if the activity belongs to current user, and trim if username is too long
+        String username = isCurrentUser ? "You" :
+                author.getUsername().length() > 12 ? author.getUsername().substring(0, 9)+"..." :
+                        author.getUsername();
+        String textContents = "";
+        switch(type){
+            case DROP:
+                textContents = username + " dropped a pin";
+                holder.icon.setImageResource(R.drawable.ic_drop);
+                break;
+            case FIND:
+                textContents = username + " found a pin";
+                holder.icon.setImageResource(R.drawable.ic_search);
+                break;
+            case COMMENT:
+                textContents = username + " commented on a pin";
+                holder.icon.setImageResource(R.drawable.ic_comment);
+                break;
+            case FOLLOW:
+                holder.icon.setImageResource(R.drawable.ic_follow);
+                User cachedPinAuthor = firebase.getCachedUser(id);
+                if (cachedPinAuthor != null) {
+                    textContents = username + " followed " + cachedPinAuthor.getUsername();
+                } else {
+                    firebase.fetchUser(id).addOnCompleteListener(task -> {
+                        String finalTextContents =
+                                username + " followed " + task.getResult().getUsername();
```

Comment:
Looks like we don't truncate the `cachedPinAuthor` username (or the result of the fetch user task). Shouldn't we truncate this too just in case their username is super long?

Reply:
Truncating the second username is not necessary since we already ellipsize the whole string if it is too long